File: android/app/src/main/java/com/plantora/billing/ui/sales/DuesViewModel.kt
package com.plantora.billing.ui.sales

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.plantora.billing.data.AuthState
import com.plantora.billing.data.BillRepository
import com.plantora.billing.data.SessionRepository
import com.plantora.billing.data.SettlementRepository
import com.plantora.billing.data.remote.friendlyError
import com.plantora.billing.domain.BillListEntry
import com.plantora.billing.domain.Money
import com.plantora.billing.domain.Role
import com.plantora.billing.domain.daysSince
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Outstanding dues across the whole shop (shared by all staff). Collecting a due
 * settles it into cash and/or UPI. A manager's collection applies immediately; a
 * salesperson's collection is sent to a manager to approve first (the bill stays
 * outstanding, flagged "Waiting for approval", until then).
 */
/** Dues unpaid for at least this many days are flagged "priority" (overdue). */
private const val PRIORITY_DAYS = 30

/** How a collection is being split. */
enum class SettleMode { CASH, UPI, SPLIT }

/** The "collect this due" bottom sheet. A partial amount may be collected — the
 *  rest stays owed. `amount` is how much to collect now (≤ the due). */
data class SettleTarget(
    val entry: BillListEntry,
    val amount: String = entry.dueAmount.toWire(),
    val mode: SettleMode = SettleMode.CASH,
    val splitCash: String = "",
    val submitting: Boolean = false,
    val error: String? = null,
) {
    /** How much is being collected now. */
    val total: Money get() = Money.parse(amount)

    /** The cash portion being collected, given the mode. */
    val cashAmount: Money
        get() = when (mode) {
            SettleMode.CASH -> total
            SettleMode.UPI -> Money.ZERO
            SettleMode.SPLIT -> Money.parse(splitCash)
        }

    /** The UPI portion being collected, given the mode. */
    val upiAmount: Money
        get() = when (mode) {
            SettleMode.CASH -> Money.ZERO
            SettleMode.UPI -> total
            SettleMode.SPLIT -> (total - Money.parse(splitCash)).let { if (it.isNegative()) Money.ZERO else it }
        }

    /** Positive, no more than what's owed, and any split fits inside the amount. */
    val valid: Boolean
        get() = total.isPositive() && total <= entry.dueAmount &&
            (mode != SettleMode.SPLIT || Money.parse(splitCash) <= total)
}

data class DuesUiState(
    val loading: Boolean = true,
    val error: String? = null,
    val dues: List<BillListEntry> = emptyList(),
    val query: String = "",
    val isManager: Boolean = false,
    val settle: SettleTarget? = null,
    val message: String? = null,
) {
    val totalOwed: Money get() = dues.fold(Money.ZERO) { acc, b -> acc + b.dueAmount }

    /** Dues matching the search box (by customer name or phone). */
    private val filtered: List<BillListEntry>
        get() {
            val q = query.trim()
            if (q.isBlank()) return dues
            return dues.filter { e ->
                e.customerName?.contains(q, ignoreCase = true) == true ||
                    e.customerPhone?.contains(q) == true
            }
        }

    /** Overdue (30+ days) dues, oldest first — shown in the priority section. */
    val priorityDues: List<BillListEntry>
        get() = filtered.filter { (daysSince(it.createdAt) ?: 0) >= PRIORITY_DAYS }
            .sortedByDescending { daysSince(it.createdAt) ?: 0 }

    /** The remaining (newer) dues. */
    val otherDues: List<BillListEntry>
        get() = filtered.filter { (daysSince(it.createdAt) ?: 0) < PRIORITY_DAYS }

    val hasResults: Boolean get() = filtered.isNotEmpty()
}

@HiltViewModel
class DuesViewModel @Inject constructor(
    private val billRepo: BillRepository,
    private val settlementRepo: SettlementRepository,
    session: SessionRepository,
) : ViewModel() {

    private val _ui = MutableStateFlow(DuesUiState())
    val ui: StateFlow<DuesUiState> = _ui.asStateFlow()

    init {
        val isManager = (session.state.value as? AuthState.Authenticated)?.user?.role == Role.MANAGER
        _ui.update { it.copy(isManager = isManager) }
        load()
    }

    fun load() {
        _ui.update { it.copy(loading = true, error = null) }
        viewModelScope.launch {
            runCatching { billRepo.listDues() }
                .onSuccess { page -> _ui.update { it.copy(loading = false, dues = page.items) } }
                .onFailure { e -> _ui.update { it.copy(loading = false, error = friendlyError(e)) } }
        }
    }

    // ── Collect (settle) sheet ──
    fun openSettle(entry: BillListEntry) = _ui.update { it.copy(settle = SettleTarget(entry)) }
    fun closeSettle() = _ui.update { it.copy(settle = null) }
    fun setSettleAmount(v: String) = _ui.update { s ->
        val t = s.settle ?: return@update s
        s.copy(settle = t.copy(amount = v, error = null))
    }

    fun setSettleMode(mode: SettleMode) = _ui.update { s ->
        // Seed the split's cash part with the whole amount so it starts valid.
        val t = s.settle ?: return@update s
        val seeded = if (mode == SettleMode.SPLIT && t.splitCash.isBlank())
            t.copy(mode = mode, splitCash = t.total.toWire(), error = null)
        else t.copy(mode = mode, error = null)
        s.copy(settle = seeded)
    }

    fun setSettleSplitCash(v: String) = _ui.update { s ->
        val t = s.settle ?: return@update s
        s.copy(settle = t.copy(splitCash = v, error = null))
    }

    fun confirmSettle() {
        val t = _ui.value.settle ?: return
        if (t.submitting) return
        if (!t.valid) {
            _ui.update { it.copy(settle = t.copy(error = "Enter an amount up to ${t.entry.dueAmount.format()}.")) }
            return
        }
        val remainingDue = (t.entry.dueAmount - t.total).let { if (it.isNegative()) Money.ZERO else it }
        _ui.update { it.copy(settle = t.copy(submitting = true, error = null)) }
        viewModelScope.launch {
            runCatching { settlementRepo.collect(t.entry.id, t.cashAmount, t.upiAmount) }
                .onSuccess { status ->
                    if (status == "approved") {
                        // Applied now. Drop it if fully cleared, else keep with the
                        // reduced due still showing.
                        _ui.update { s ->
                            s.copy(
                                settle = null,
                                dues = if (remainingDue.isPositive())
                                    s.dues.map { if (it.id == t.entry.id) it.copy(dueAmount = remainingDue) else it }
                                else s.dues.filterNot { it.id == t.entry.id },
                                message = if (remainingDue.isPositive())
                                    "Collected ${t.total.format()} — ${remainingDue.format()} still owed."
                                else "Collected — ${t.total.format()}.",
                            )
                        }
                    } else {
                        // Pending — keep it in the list but flag it so it can't be
                        // collected again while a manager reviews it.
                        _ui.update { s ->
                            s.copy(
                                settle = null,
                                dues = s.dues.map { if (it.id == t.entry.id) it.copy(pendingSettlement = true) else it },
                                message = "Sent to the manager for approval.",
                            )
                        }
                    }
                }
                .onFailure { e ->
                    _ui.update { it.copy(settle = it.settle?.copy(submitting = false, error = friendlyError(e, "Couldn't record the collection."))) }
                }
        }
    }

    fun setQuery(q: String) = _ui.update { it.copy(query = q) }

    fun dismissMessage() = _ui.update { it.copy(message = null) }
}

File: Plantbill/Plantbill/Features/Sales/DuesViewModel.swift
import Foundation
import Observation

/// Dues unpaid for at least this many days are flagged "priority" (overdue).
private let priorityDays = 30

enum SettleMode: String {
    case cash, upi, split
}

/// The "collect this due" sheet state. A partial amount may be collected —
/// the rest stays owed. `amount` is how much to collect now (≤ the due).
struct SettleTarget: Equatable {
    let entry: BillListEntry
    var amount: String
    var mode: SettleMode = .cash
    var splitCash: String = ""
    var submitting = false
    var error: String?

    init(entry: BillListEntry) {
        self.entry = entry
        self.amount = entry.dueAmountMoney.toInput()
    }

    var total: Money { Money.parse(amount) }

    var cashAmount: Money {
        switch mode {
        case .cash: return total
        case .upi: return .zero
        case .split: return Money.parse(splitCash)
        }
    }

    var upiAmount: Money {
        switch mode {
        case .cash: return .zero
        case .upi: return total
        case .split:
            let remainder = total - Money.parse(splitCash)
            return remainder.isNegative ? .zero : remainder
        }
    }

    var valid: Bool {
        total.isPositive && total <= entry.dueAmountMoney && (mode != .split || Money.parse(splitCash) <= total)
    }
}

@Observable
@MainActor
final class DuesViewModel {
    let isManager: Bool

    private(set) var isLoading = true
    private(set) var loadError: String?
    private(set) var dues: [BillListEntry] = []
    var query: String = ""
    var settle: SettleTarget?
    var message: String?

    init(isManager: Bool) {
        self.isManager = isManager
    }

    var totalOwed: Money { dues.reduce(Money.zero) { $0 + $1.dueAmountMoney } }

    private var filtered: [BillListEntry] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return dues }
        return dues.filter {
            ($0.customerName?.localizedCaseInsensitiveContains(trimmed) ?? false)
                || ($0.customerPhone?.contains(trimmed) ?? false)
        }
    }

    var priorityDues: [BillListEntry] {
        filtered.filter { $0.daysSinceCreated >= priorityDays }.sorted { $0.daysSinceCreated > $1.daysSinceCreated }
    }

    var otherDues: [BillListEntry] {
        filtered.filter { $0.daysSinceCreated < priorityDays }
    }

    var hasResults: Bool { !filtered.isEmpty }

    func load() async {
        isLoading = true
        loadError = nil
        do {
            let query = [
                URLQueryItem(name: "has_due", value: "true"),
                URLQueryItem(name: "limit", value: "100"),
            ]
            let page: BillListPage = try await APIClient.shared.send(Endpoint(path: "bills", queryItems: query))
            dues = page.items.filter { $0.dueAmountMoney.isPositive }
            isLoading = false
        } catch let error as APIError {
            isLoading = false
            loadError = error.userMessage
        } catch {
            isLoading = false
            loadError = APIError.unknown.userMessage
        }
    }

    // MARK: Settle sheet

    func openSettle(_ entry: BillListEntry) { settle = SettleTarget(entry: entry) }
    func closeSettle() { settle = nil }

    func setSettleMode(_ mode: SettleMode) {
        guard var target = settle else { return }
        if mode == .split && target.splitCash.isEmpty {
            target.splitCash = target.total.toInput()
        }
        target.mode = mode
        target.error = nil
        settle = target
    }

    func confirmSettle() async {
        guard var target = settle, !target.submitting else { return }
        guard target.valid else {
            target.error = "Enter an amount up to what's owed (\(target.entry.dueAmountMoney.format()))."
            settle = target
            return
        }
        target.submitting = true
        target.error = nil
        settle = target

        do {
            let request = SettlementCreateRequest(billId: target.entry.id, cashAmount: target.cashAmount.toWire(), upiAmount: target.upiAmount.toWire())
            let body = try APIClient.shared.encode(request)
            let result: SettlementActionResult = try await APIClient.shared.send(Endpoint(path: "settlements", method: .post, body: body))
            let remainingDue = target.entry.dueAmountMoney - target.total
            let remaining = remainingDue.isNegative ? Money.zero : remainingDue

            if result.status == "approved" {
                if remaining.isPositive {
                    if let index = dues.firstIndex(where: { $0.id == target.entry.id }) {
                        dues[index] = replacingDue(dues[index], due: remaining)
                    }
                    message = "Collected \(target.total.format()) — \(remaining.format()) still owed."
                } else {
                    dues.removeAll { $0.id == target.entry.id }
                    message = "Collected \(target.total.format())."
                }
            } else {
                if let index = dues.firstIndex(where: { $0.id == target.entry.id }) {
                    dues[index] = replacingPending(dues[index])
                }
                message = "Sent to your manager for approval."
            }
            settle = nil
        } catch let error as APIError {
            target.submitting = false
            target.error = error.userMessage
            settle = target
        } catch {
            target.submitting = false
            target.error = APIError.unknown.userMessage
            settle = target
        }
    }

    private func replacingDue(_ entry: BillListEntry, due: Money) -> BillListEntry {
        BillListEntry(id: entry.id, createdAt: entry.createdAt, total: entry.total, dueAmount: due.toWire(), customerName: entry.customerName, customerPhone: entry.customerPhone, itemCount: entry.itemCount, paymentMethod: entry.paymentMethod, isEdited: entry.isEdited, pendingSettlement: entry.pendingSettlement)
    }

    private func replacingPending(_ entry: BillListEntry) -> BillListEntry {
        BillListEntry(id: entry.id, createdAt: entry.createdAt, total: entry.total, dueAmount: entry.dueAmount, customerName: entry.customerName, customerPhone: entry.customerPhone, itemCount: entry.itemCount, paymentMethod: entry.paymentMethod, isEdited: entry.isEdited, pendingSettlement: true)
    }

    func dismissMessage() { message = nil }
}
